docs/adr/0002-delivery-channel-creation-and-integration.md

```diff
@@ -0,0 +1,102 @@
+         v           v                     v
++--------+-+      +--+-------+           +-+--------+
+|          |      |          |           |          |
+| Delivery |      | Delivery |           | Delivery |
+| Channel  |      | Channel  | +-+  +-+  | Channel  |
+| 1        |      | 2        |           | N        |
+|          |      |          |           |          |
+|          |      |          |           |          |
++----------+      +----------+           +----------+
+```
+
+The management api should have knowledge of what delivery channels are available and where certain notification types should be sent. 
+
+When a new notification for a user is sent to the management api, the management api will read the user preference for the said notification type and see which communication types the user has 'subscribed' to. Following this, the notification delivered through the appropriate delivery channel.
+
+### User notifications and user settings
+
+A user should be able to choose what kind of communication types they receive based on the notification type. There are two ways of approaching this, either:
+1. Send different types of notifications through all channels that are provided by the consumer of this service, and leave the communication type and user configuration for the consumer of the Marain.UserNotifications service. (ie, the consumer of the Marain.UserNotifications service should specify which delivery channels they want to send notifications via an api call when a notification is created)
+2. Store user preferences somewhere that is accessible to the management api, then when a new notification with a certain notification type comes through, check what type of communication types  the user has 'subscribed' to and then send a message through the delivery channel.
+
+The latter option was chosen as it provides a user the ability to choose what notifications they want out of the box, if the former is chosen, all users/businesses integrating with this service will then need to write their own notifications user management service.
+
+
+
+Settings for which communication channel a user can receive a notification through will be stored inside a User Notifications Settings table which will be accessed by the Management Api. Configuration in this table will include the Notification Type and the appropriate communication types associated to that notification type. This table will have an associated crud api that will allow a consumer of `Marain.UserNotifications` to view/change the stored settings.
+
+### Management Api and usable Delivery Channels
+
+The management API will host configuration per tenant for what communication types can be sent through which delivery channels in the following format:
+
+```json
+[
+  {
+    "id": "dc1",
+    "displayName": "Airship",
+    "communicationTypes": ["email", "sms", "web-push"]
+  }, 
+  {
+    "id": "dc2",
+    "displayName": "Twilio",
+    "communicationTypes": ["sms"]
+  }
+]
```

Comment:
Is this realistic? If this is configuration for the tenant, wouldn't it only specify a single delivery channel for each communication type? What does it mean for a tenant to specify that SMSs should be delivered by both the Airship and Twilio channels? Wouldn't that mean SMS notifications would be delivered twice?

Also, I think I've seen some updates to this structure in calls we've had, showing how the channel-specific configuration (endpoints, keyvault secret references, etc) would be added to this config, so this needs an update.

Reply:
This is an interesting question. If someone onboards their tenant with Airship and Twilio, then all of those channels are configurable and available. Which ones they choose to use in a particular application, and then which they choose to use for a particular notification is a different question again, isn't it?

Reply:
I think that
- the user configuration will associate communication type (WebPush, SMS, etc) with notification type (with the option of receiving notifications via multiple communication types if you so desire - another possible future enhancement might be a new communication type for "daily summary by email", for example, so you might choose to receive notifications by SMS as they are raised, and also receive a daily summary)
- the tenant configuration will associate each communication type with a single delivery channel - so even if you have configuration for Airship and Twilio, you would specify that all SMS messages will go via Twilio, regardless of the fact that Airship is capable of sending them.

I'd prefer to keep the bits of code in the consuming system that create notifications from having to know anything about how they are ultimately delivered. And obviously the user should have no visibility of the underlying delivery channels - they should only care about communication types.

That said, I guess you could (for example) consider having "primary" and "fallback" delivery channels for each communication type,, but I think that's beyond the scope of what we're doing right now (and relatively simple to add in future if we ever decide we want it).

Reply:
We have removed this section as it looks like the previous section already details how this will work. We have made updates to this. I will go over the full document and add/remove bits that have changed over the last month.feat

@mwadamsdentsu, At the moment we're not planning on sending certain notification to a delivery channel based on their notification type, only communication type.

Reply:
Yes, agreed Jon. That is the way that we've currently designed it. The delivery channels are only configured when the Marain.UserNotifications is set up.

I also agree that having a "primary" and "fallback" delivery channel is out of the scope of this ADR.